src/main/java/org/jenkinsci/plugins/workflow/graph/FlowNode.java

```diff
@@ -109,14 +109,39 @@ protected Object readResolve() throws ObjectStreamException {
      * <p>It will be false for a node which still has active children, like a step with a running body.
      * It will also be false for something that has finished but is pending child node creation,
      * such as a completed fork branch which is waiting for the join node to be created.
-     * <p>This can only go from true to false and is a shortcut for {@link FlowExecution#isCurrentHead}.
+     * <p>This can only go from true to false.
+     * @deprecated Usually {@link #isActive} is what you want. If you really wanted the original behavior, use {@link FlowExecution#isCurrentHead}.
      */
-    @Exported
+    @Deprecated
     public final boolean isRunning() {
         return getExecution().isCurrentHead(this);
     }
 
     /**
+     * Checks whether a node is still part of the active part of the graph.
+     * Unlike {@link #isRunning}, this behaves intuitively for a {@link BlockStartNode}:
+     * it will be considered active until the {@link BlockEndNode} is added.
+     */
+    @Exported(name="running")
+    public final boolean isActive() {
+        if (this instanceof FlowEndNode) { // cf. JENKINS-26139
+            return false;
+        }
+        List<FlowNode> currentHeads = exec.getCurrentHeads();
+        if (currentHeads.contains(this)) {
+            return true;
+        }
+        if (this instanceof BlockStartNode) {
+            for (FlowNode headNode : currentHeads) {
+                if (new LinearBlockHoppingScanner().findFirstMatch(headNode, Predicates.equalTo(this)) != null) {
```

Comment:
@jglick 🐛  If you use this implementation you're potentially scanning all the nodes before a parallel multiple times, remember?   So if you have this: 

```
stage ('some long stuff') {
  for(int i=0; i<1000; i++) {
     echo("say something for the $i-th time")
  }
}
parallel withNBranches
```
then you're scanning 'some long stuff' up to N times for N*1000+4 nodes. 

In *this* specific testcase you're seeing faster build times but you're optimizing just for that case.  

Better solution: go back to the cache-like method, but make the cache persisted within the FlowExecution or something like that (and rebuild it if not present).  

Long term we need to support this within either the `FlowNodeStorage` or provide a lookup object with graph structure information that we can use to accelerate lookups (AKA Bismuth Level 3, the tree-structured view of blocks).